net: add ASMap info in `getrawaddrman` RPC

brunoerg opened the pull request:
This PR adds two new fields in `getrawaddrman` RPC: "mapped_as" and "source_mapped_as". These fields are used to return the ASN (Autonomous System Number) mapped to the peer and its source. With these informations we can have a better view of the bucketing logic with ASMap specially in projects like [addrman-observer](https://github.com/0xb10c/addrman-observer).

brunoerg committed:
test: add coverage for `mapped_as` from `getrawaddrman`

Test addresses are being mapped according to the ASMap
file provided properly. Compare the result of the `getrawaddrman`
RPC with the result from the ASMap Health Check.

brunoerg committed:
net: rpc: return peer's mapped AS in getrawaddrman

This information can be used to check bucketing
logic.

brunoerg commented:
ping @0xB10C

0xB10C commented:
Cool! Concept ACK.

I've added prelimarly support for this in the static HTML page on https://addrman.observer. You can now load the new `getrawaddrman` dumps and it will show you the AS and source AS. You can also mouse-over highlight by AS and source ASN.

![image](https://github.com/bitcoin/bitcoin/assets/19157360/e5556e13-6c03-4183-bdda-15eedb7c8397)

0xB10C commented on src/rpc/net.cpp:
```diff
@@ -1103,10 +1103,12 @@ UniValue AddrmanEntryToJSON(const AddrInfo& info)
     ret.pushKV("network", GetNetworkName(info.GetNetClass()));
     ret.pushKV("source", info.source.ToStringAddr());
     ret.pushKV("source_network", GetNetworkName(info.source.GetNetClass()));
+    ret.pushKV("mapped_as", connman.GetMappedAS(info));
+    ret.pushKV("source_mapped_as", connman.GetMappedAS(info.source));
```
nit: I always found this confusing as I read `mapped as <something>` and not `mapped *A*utonomous*S*ystem` - don't have a concrete idea for a name and I guess it makes sense to use the same as in `getpeerinfo`.

brunoerg replied:
I wanted to keep same name as in `getpeerinfo` but I agree with you about the confusion, I do believe `mapped_asn` would be better because we show the AS number.

brunoerg replied:
I will leave as is for now.

brunoerg commented:
cc: @virtu @fjahr

virtu commented on src/rpc/net.cpp:
```diff
@@ -1103,10 +1103,12 @@ UniValue AddrmanEntryToJSON(const AddrInfo& info)
     ret.pushKV("network", GetNetworkName(info.GetNetClass()));
     ret.pushKV("source", info.source.ToStringAddr());
     ret.pushKV("source_network", GetNetworkName(info.source.GetNetClass()));
+    ret.pushKV("mapped_as", connman.GetMappedAS(info));
+    ret.pushKV("source_mapped_as", connman.GetMappedAS(info.source));
```
nit: Looking at my node's `getrawaddrman` output, it felt odd to have `mapped_as` appear after `source`. IMO, the fields become more intuitively understandable when put right after or at least close to the address they refer to.

brunoerg replied:
I agree, will address it.

virtu reviewed (commented):
Concept ACK

Also did some light testing:
- Some sanity checks on `getrawaddrman` output (everything mapped to zero when asmap is disabled; almost all addresses mapped to non-zero when enabled)
- Functional tests

fjahr reviewed (commented):
Concept ACK

fjahr commented on src/rpc/net.cpp:
```diff
@@ -1139,6 +1141,8 @@ static RPCHelpMan getrawaddrman()
                         {RPCResult::Type::NUM_TIME, "time", "The " + UNIX_EPOCH_TIME + " when the node was last seen"},
                         {RPCResult::Type::STR, "source", "The address that relayed the address to us"},
                         {RPCResult::Type::STR, "source_network", "The network (" + Join(GetNetworkNames(), ", ") + ") of the source address"},
+                        {RPCResult::Type::NUM, "mapped_as", "The AS in the BGP route mapped to the node (0 if not mapped)"},
```
> I would prefer these to be optional results rather than a magic 0 value.

(if there is no mapping of course)

brunoerg replied:
I agree, I will address it.

brunoerg replied:
Done

brunoerg commented:
Force-pushed:

- Changed `mapped_as` and `source_mapped_as` to optional.
- Changed description of the fields as suggested by @fjahr.
- Reordered these new fields (https://github.com/bitcoin/bitcoin/pull/30062#discussion_r1607941908)

fjahr commented on src/rpc/net.cpp:
```diff
@@ -1133,12 +1141,14 @@ static RPCHelpMan getrawaddrman()
                         {RPCResult::Type::NUM_TIME, "time", "The " + UNIX_EPOCH_TIME + " when the node was last seen"},
                         {RPCResult::Type::STR, "source", "The address that relayed the address to us"},
                         {RPCResult::Type::STR, "source_network", "The network (" + Join(GetNetworkNames(), ", ") + ") of the source address"},
+                        {RPCResult::Type::NUM, "source_mapped_as", /*optional=*/true, "The ASN mapped to the IP of this peer's source per our current ASMap."}
```
nit: This comment has a dot at the end while our convention seems to be to not add a dot.

brunoerg replied:
done

fjahr commented on test/functional/feature_asmap.py:
```diff
@@ -26,6 +26,7 @@
 import os
 import shutil
 
+from test_framework.util import assert_equal
```
nit: By alphabetical ordering this would go below the following line

brunoerg replied:
done

fjahr reviewed (commented):
Code review ACK 2de765c25017e566777d927d86452d96082592a1

Feel free to ignore the nits but I can re-review quickly if you address them.

fjahr commented:
Code review ACK 1e54d61c4698debf3329d1960e06078ccbf8063c

0xB10C reviewed (approved):
ACK 1e54d61c4698debf3329d1960e06078ccbf8063c

I've played around with the `getrawaddrman` RPC with and without an asmap file loaded. Also loaded a json dump into addrman-observer. Works as expected.

virtu commented:
ACK [1e54d61](https://github.com/bitcoin/bitcoin/commit/1e54d61c4698debf3329d1960e06078ccbf8063c)

Successfully re-ran all previously mentioned tests.

glozow commented on test/functional/feature_asmap.py:
```diff
@@ -118,6 +119,14 @@ def test_asmap_health_check(self):
         with self.node.assert_debug_log(expected_msgs=[msg]):
             self.start_node(0, extra_args=['-asmap'])
+        raw_addrman = self.node.getrawaddrman()
+        asns = []
```
nit: this could have been a set?

glozow reviewed (commented):
ACK 1e54d61c4698debf3329d1960e06078ccbf8063c

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -1093,20 +1093,28 @@ static RPCHelpMan getaddrmaninfo()
 }
 
-UniValue AddrmanEntryToJSON(const AddrInfo& info)
+UniValue AddrmanEntryToJSON(const AddrInfo& info, CConnman& connman)
```
Makes sense, I will address it in a follow-up PR.

brunoerg commented:
Thanks for reviewing, @jonatack. I'm working on a follow-up.
